src/OpenRiaServices.DomainServices.Hosting/Framework/Services/MessageEncoders/PoxBinaryMessageEncodingBindingElement.cs

```diff
@@ -255,15 +256,224 @@ public override ArraySegment<byte> WriteMessage(Message message, int maxMessageS
+
+                public override long Seek(long offset, SeekOrigin origin)
+                {
+                    throw new NotImplementedException();
+                }
+
+                public override void SetLength(long value)
+                {
+                    throw new NotImplementedException();
+                }
+
+                public override void Write(byte[] buffer, int offset, int count)
+                {
+                    // Argument validation is skipped since it is only used by 
+                    // BinaryXml writer which we trust to always give valid input
 
-                    byte[] buffer = bufferManager.TakeBuffer((int)ms.Position + messageOffset);
-                    Buffer.BlockCopy(ms.GetBuffer(), 0, buffer, messageOffset, (int)ms.Position);
-                    return new ArraySegment<byte>(buffer, messageOffset, (int)ms.Position);
+                    // Note: BinaryXml buffers up to 512 bytesso we should expect most writes to be around 
+                    // 500+ bytes (smaller if the next write is a long string or byte array)
+
+                    do
+                    {
+                        EnsureBufferCapacity();
+
+                        // Write bufffer
+                        if (count <= _buffer.Length - _bufferWritten)
+                        {
+                            FastCopy(buffer, offset, _buffer, _bufferWritten, count);
+                            _position += count;
+                            _bufferWritten += count;
+                            break;
+                        }
+                        else
+                        {
+                            // Fill _buffer
+                            int toCopy = _buffer.Length - _bufferWritten;
+                            FastCopy(buffer, offset, _buffer, _bufferWritten, toCopy);
+                            _position += toCopy;
+                            offset += toCopy;
+                            count -= toCopy;
+                        }
+                    } while (count > 0);
+                }
+
+                /// <summary>
+                /// Allocate more space if buffer is full.
+                /// Ensures _buffer is non null and has space to write more bytes
+                /// </summary>
+                private void EnsureBufferCapacity()
+                {
+                    // There is space left
+                    if (_bufferWritten < _buffer.Length)
+                        return;
+
+                    // Save current buffer in list before allocating a new buffer
+                    if (_bufferList == null)
+                        _bufferList = new System.Collections.Generic.List<byte[]>(capacity: 16);
+                    _bufferList.Add(_buffer);
+                    // Ensure we never return buffer twice in case TakeBuffer below throws
+                    _buffer = null;
+
+                    _buffer = _bufferManager.TakeBuffer(Math.Min(_position, _maxSize));
```

Comment:
Consider using position ×2

+ can maybe return the last buffer (to offset increase space usage)
+ less different sizes should make the buffermanager more efficient